api/src/main/java/io/opentelemetry/trace/SpanId.java

```diff
@@ -81,45 +60,19 @@ public static SpanId getInvalid() {
    * @param random The random number generator.
    * @return a valid new {@code SpanId}.
    */
-  static SpanId generateRandomId(Random random) {
+  static CharSequence generateRandomId(Random random) {
     long id;
     do {
       id = random.nextLong();
     } while (id == INVALID_ID);
-    return new SpanId(id);
+    return fromLong(id);
   }
 
-  /**
-   * Returns a {@code SpanId} whose representation is copied from the {@code src} beginning at the
-   * {@code srcOffset} offset.
-   *
-   * @param src the buffer where the representation of the {@code SpanId} is copied.
-   * @param srcOffset the offset in the buffer where the representation of the {@code SpanId}
-   *     begins.
-   * @return a {@code SpanId} whose representation is copied from the buffer.
-   * @throws NullPointerException if {@code src} is null.
-   * @throws IndexOutOfBoundsException if {@code srcOffset+SpanId.getSize()} is greater than {@code
-   *     src.length}.
-   * @since 0.1.0
-   */
-  public static SpanId fromBytes(byte[] src, int srcOffset) {
-    Utils.checkNotNull(src, "src");
-    return new SpanId(BigendianEncoding.longFromByteArray(src, srcOffset));
-  }
-
-  /**
-   * Copies the byte array representations of the {@code SpanId} into the {@code dest} beginning at
-   * the {@code destOffset} offset.
-   *
-   * @param dest the destination buffer.
-   * @param destOffset the starting offset in the destination buffer.
-   * @throws NullPointerException if {@code dest} is null.
-   * @throws IndexOutOfBoundsException if {@code destOffset+SpanId.getSize()} is greater than {@code
-   *     dest.length}.
-   * @since 0.1.0
-   */
-  public void copyBytesTo(byte[] dest, int destOffset) {
-    BigendianEncoding.longToByteArray(id, dest, destOffset);
+  /** Converts the long id value into a base-16 representation of it. */
+  public static CharSequence fromLong(long id) {
+    char[] chars = new char[BASE16_SIZE];
```

Comment:
I think you can get a significant improvement by using a recycled buffer, maybe closing the gap you're seeing. Here's an example

https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/internal/RecyclableBuffers.java

It's a bit confusing since it used to be sized to just fit an ID but also got generalized to use when generating trace headers, which would probably be nice to do here too in a separate PR :)

Reply:
definitely a good idea!